### PR TITLE
deprecate hpc_load() and integrate it with restore()

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

Part of #7652

`CheckpointConnector.hpc_load()` does the same thing as `CheckpointConnector()`, except it additionally calls a hook
on_hpc_load on the module. This hook calling is now taken care of in `restore()`.
Now we can properly deprecate the hpc_load for `restore()` to remove the code duplication.

The unit tests added here ensure that `restore()` behaves the same as the deprecated `hpc_load`!

Also closes #5371 

## Before submitting
- [ ] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/trainer/connectors/checkpoint_connector.py [hunk 1]
```diff
@@ -248,6 +257,20 @@ def restore_lr_schedulers(self) -> None:
     # ----------------------------------
     # PRIVATE OPS
     # ----------------------------------
```
Comment (carmocca) [inline pytorch_lightning/trainer/connectors/checkpoint_connector.py:259]: The function is under `private ops` yet we are deprecating it.

Are there any places where we recommend overriding it? Or any docs or public example which uses it?

File: pytorch_lightning/trainer/connectors/checkpoint_connector.py [hunk 2]
```diff
@@ -347,6 +370,7 @@ def dump_checkpoint(self, weights_only: bool = False) -> dict:
 
             self.trainer.precision_plugin.on_save_checkpoint(checkpoint)
 
+        # dump hyper-parameters
```
Comment (carmocca) [inline pytorch_lightning/trainer/connectors/checkpoint_connector.py:373]: dupe

```suggestion
```

File: pytorch_lightning/trainer/connectors/checkpoint_connector.py [hunk 3]
```diff
@@ -365,30 +389,6 @@ def dump_checkpoint(self, weights_only: bool = False) -> dict:
 
         return checkpoint
 
-    def hpc_load(self, checkpoint_path: str, on_gpu: bool):
```
Comment (carmocca) [inline pytorch_lightning/trainer/connectors/checkpoint_connector.py:368]: nit:

you could put the new function where it was so the diff is smaller and history is better preserved

File: tests/trainer/connectors/test_checkpoint_connector.py [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+        checkpoint_callback=False,
+    )
+    trainer.fit(model)
+    connector = trainer.checkpoint_connector
+    connector.hpc_save(tmpdir, logger=trainer.logger)
+    assert model.hpc_save_called == 1
+    assert model.hpc_load_called == 0
+
+    # new training run, restore from hpc checkpoint file automatically
+    assert set(os.listdir(tmpdir)) == {"hpc_ckpt_1.ckpt", "lightning_logs"}
```
Comment (carmocca) [inline tests/trainer/connectors/test_checkpoint_connector.py:52]: set `logger=False` so `lightning_logs` isn't there?

File: pytorch_lightning/trainer/connectors/checkpoint_connector.py (hunk 1 above)
Comment (awaelchli) [reply pytorch_lightning/trainer/connectors/checkpoint_connector.py:259]: This comment was already there. If I don't deprecate someone is going to comment about it because it was a public method. It does not hurt

File: tests/trainer/connectors/test_checkpoint_connector.py (hunk 4 above)
Comment (awaelchli) [reply tests/trainer/connectors/test_checkpoint_connector.py]: sure, I can give a mock to hpc_save instead :)

File: pytorch_lightning/trainer/connectors/checkpoint_connector.py (hunk 3 above)
Comment (awaelchli) [reply pytorch_lightning/trainer/connectors/checkpoint_connector.py:368]: what if the intention was to move it? 

done :)

**Review by justusschock — APPROVED** [review]

**Review by carmocca — APPROVED** [review]

**Review by Borda — APPROVED** [review]
I may use this wrapper but otherwise it looks food to me https://github.com/Borda/pyDeprecate#simple-function-forwarding
